### PR TITLE
feat: add drive multi-selection in store

### PR DESCRIPTION
**Pull request by Shou**
We lay the foundation for multi-selecting drives by implementing it into
the `store` and relevant modules interacting with the `store`.

Changelog-Entry: Add drive multi-selection to the store.

### CODE REVIEW

**Comment by Shou** [comment]
I've rebased this with some fixes to pass the tests.

File: tests/gui/pages/main.spec.js [hunk 1]
```diff
@@ -154,6 +154,9 @@ describe('Browser: MainPage', function () {
           }
         })
 
+        console.log(`hasDrive: ${selectionState.hasDrive()}`)
+        console.log(`hasImage: ${selectionState.hasImage()}`)
```
Comment (lurch) [inline tests/gui/pages/main.spec.js]: Is this leftover debugging code?

File: tests/shared/models/selection-state.spec.js [hunk 2]
```diff
@@ -129,16 +131,101 @@ describe('Model: selectionState', function () {
...
+    })
+
+    it('should be able to remove a drive', function () {
+      m.chai.expect(selectionState.getSelectedDevices().length).to.equal(2)
+      selectionState.toggleDrive(this.drives[0].device)
+      m.chai.expect(selectionState.hasDrive()).to.be.true
+      m.chai.expect(selectionState.getSelectedDevices().length).to.equal(1)
+      m.chai.expect(selectionState.getSelectedDevices()).to.deep.equal([ this.drives[1].device ])
+    })
+  })
```
Comment (lurch) [inline tests/shared/models/selection-state.spec.js]: Some of these unit-tests feel like a bit of a mish-mash? Could you come up with a neat & logical set of more thorough unit-tests?

File: tests/shared/models/selection-state.spec.js [hunk 3]
```diff
@@ -743,7 +830,7 @@ describe('Model: selectionState', function () {
         selectionState.removeImage()
       })
 
-      it('should de-select a previously selected not-recommended drive', function () {
+      it('should not de-select a previously selected not-recommended drive', function () {
```
Comment (lurch) [inline tests/shared/models/selection-state.spec.js]: Is this a behaviour change, or was the test-description previously incorrect?

File: lib/shared/models/selection-state.js [hunk 4]
```diff
@@ -75,21 +75,54 @@ exports.setImage = (image) => {
...
+exports.getDevices = () => {
+  return store.getState().getIn([ 'selection', 'drives' ]).toJS()
+}
+
+/**
+ * @summary Get all selected drive objects
+ * @function
+ * @public
+ *
+ * @returns {String[]} selected devices
```
Comment (lurch) [inline lib/shared/models/selection-state.js]: Hmm, the `@summary` says `selected drive objects`, but the `@returns` says a String array of `selected devices`??

I see that the old `.getDrive()` function returned a drive Object, which seems a bit inconsistent naming-wise, given that all the other Drive-related functions just accept a String parameter. So perhaps for clarity, it'd be worth having a `.getSelectedDrives()` function which returns an array of Strings, and a `.getSelectedDriveObjects()` function (if it's indeed needed) which returns an array of drive Objects?

File: lib/shared/models/selection-state.js [hunk 5]
```diff
@@ -308,33 +344,54 @@ exports.removeImage = () => {
...
-    store.dispatch({
-      type: store.Actions.REMOVE_IMAGE
-    })
+    exports.removeImage()
   }
 
-  store.dispatch({
-    type: store.Actions.REMOVE_DRIVE
-  })
+  _.map(exports.getSelectedDevices(), exports.removeDrive)
```
Comment (lurch) [inline lib/shared/models/selection-state.js]: Would it be worth having a `.removeAllDrives()` function?

File: lib/shared/models/selection-state.js [hunk 6]
```diff
@@ -21,37 +21,37 @@ const store = require('../store')
...
+ * @param {String} driveDevice - drive device identifier
  *
  * @example
- * selectionState.setDrive('/dev/disk2');
+ * selectionState.addDrive('/dev/disk2');
  */
-exports.setDrive = (drive) => {
+exports.addDrive = (driveDevice) => {
   store.dispatch({
     type: store.Actions.SELECT_DRIVE,
```
[the remark this answers is not in the document]
Comment (lurch) [reply lib/shared/models/selection-state.js]: Alternatively, we could rename `store.Actions.REMOVE_DRIVE` to `store.Actions.UNSELECT_DRIVE` ;-)

File: lib/shared/store.js [hunk 7]
```diff
@@ -151,19 +153,27 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
...
-      }
-
-      const selectedDevice = newState.getIn([ 'selection', 'drive' ])
 
-      if (selectedDevice && !_.find(action.data, {
-        device: selectedDevice
-      })) {
         return storeReducer(newState, {
-          type: ACTIONS.REMOVE_DRIVE
+          type: ACTIONS.REMOVE_DRIVE,
```
Comment (lurch) [inline lib/shared/store.js]: I still can't figure out why you've got REMOVE_DRIVE both here and on line 170 :-/

File: lib/shared/store.js [hunk 8]
```diff
@@ -414,24 +426,37 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
         })
       }
 
-      const selectedDrive = findDrive(state, state.getIn([ 'selection', 'drive' ]))
+      const selectedDevices = state.getIn([ 'selection', 'drives' ])
```
Comment (lurch) [inline lib/shared/store.js]: NIT: should this variable be named `selectedDrives` for consistency?

File: lib/shared/store.js [hunk 9]
```diff
@@ -414,24 +426,37 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
...
+          return storeReducer(accState, {
+            type: ACTIONS.REMOVE_DRIVE,
+            data: driveDevice
           })
         }
 
-        return state
-      }).setIn([ 'selection', 'image' ], Immutable.fromJS(action.data))
+        return accState
+      }, state).setIn([ 'selection', 'image' ], Immutable.fromJS(action.data))
```
Comment (lurch) [inline lib/shared/store.js]: So it looks like this e.g. removes every selected drive that is now smaller than the newly-selected image? There should definitely be unit-tests for this ;-)

File: lib/shared/store.js [hunk 10]
```diff
@@ -308,7 +318,9 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
         })
       }
 
-      return state.setIn([ 'selection', 'drive' ], Immutable.fromJS(action.data))
+      const stateDrives = state.getIn([ 'selection', 'drives' ])
```
Comment (lurch) [inline lib/shared/store.js]: Perhaps this variable should be renamed to `selectionDrives` for consistency with line 457?

File: lib/shared/models/selection-state.js [hunk 11]
```diff
@@ -75,21 +75,54 @@ exports.setImage = (image) => {
...
+ * @summary Get the head of the list of selected drives
+ * @function
+ * @public
+ *
+ * @returns {Object} drive
+ *
+ * @example
+ * const drive = selectionState.getCurrentDrive();
+ */
+exports.getCurrentDrive = () => {
```
[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/models/selection-state.js]: Basically it's at the head of the list. We need this – for now – to keep behaviour the same, applying a single drive to be flashed and not a list of drives. It's just a convenience function until we get multi-writing for real.

File: lib/shared/models/selection-state.js [hunk 12]
```diff
@@ -308,33 +344,54 @@ exports.removeImage = () => {
...
+ * }
+ */
+exports.isSelectedDrive = (driveDevice) => {
+  if (!driveDevice) {
     return false
   }
 
-  return drive === _.get(exports.getDrive(), [ 'device' ])
+  const selectedDriveDevices = exports.getSelectedDevices()
+  return Boolean(_.find(selectedDriveDevices, _.partial(_.isEqual, driveDevice)))
```
[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/models/selection-state.js]: I copied what worked elsewhere and just wrapped it in Boolean but yours is a better, more concise way, thanks!

File: lib/shared/store.js [hunk 13]
```diff
@@ -36,7 +36,9 @@ const packageJSON = require('../../package.json')
  */
 const DEFAULT_STATE = Immutable.fromJS({
   availableDrives: [],
-  selection: {},
+  selection: {
+    drives: new Immutable.OrderedSet()
```
[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/store.js]: For now it needs to be ordered, and maybe we'd prefer/require ordered sets later. This is due to it pretending to be single-selection while we're getting multi-writes in fully.

File: tests/gui/pages/main.spec.js (hunk 1 above)
Comment (Shou) [reply tests/gui/pages/main.spec.js]: Yup!

File: lib/shared/store.js (hunk 8 above)
Comment (Shou) [reply lib/shared/store.js]: Since it returns a `String[]` containing the `device` fields instead of drive objects, I think it makes more sense with `selectedDevices`.

Comment (lurch) [reply lib/shared/store.js]: For clarity, perhaps we need an Etcher-wide definition of what's a `device` or a `drive` or a `drive object` or a `driveDevice` etc., and then (try to) use those terms consistently throughout the codebase? Ping @jhermsmeier

File: lib/shared/models/selection-state.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (lurch) [reply lib/shared/models/selection-state.js]: Hmmm, so does that effectively make this a "WIP PR" until we actually have multi-writing for real? (if so, could you add the "WIP" prefix to the PR title?)

IMHO it makes little sense to merge functionality allowing a user to select multiple drives, while still only allowing the **first** drive they selected to be written to.

File: lib/shared/store.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (lurch) [reply lib/shared/store.js]: > maybe we'd prefer/require ordered sets later

The reason I'd be :-1: on an OrderedSet is because the user has no way (without unselecting everything again) of changing that ordering.

File: lib/shared/models/selection-state.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/models/selection-state.js]: It's not a WIP PR, more like a transitionary PR towards multi-writes. We can't allow users to select multiple drives when we don't actually allow flashing those multiple drives, if that makes sense.

File: lib/shared/store.js (hunk 8 above)
Comment (Shou) [reply lib/shared/store.js]: With this PR I've tried to keep `device : String` and `drive : Object`, as a starting point

File: lib/shared/models/selection-state.js (hunk 5 above)
Comment (Shou) [reply lib/shared/models/selection-state.js]: I recall using `selectionState.clear({ preserveImage: true })` as a `.removeAllDrives()` so good question, I think we should add the function and replace those `.clear()`s for clarity.

File: lib/shared/store.js (hunk 8 above)
Comment (lurch) [reply lib/shared/store.js]: But it'd still be good to nail down exactly what `device : String` and `drive : Object` actually mean, with examples of what they might contain in different scenarios, on different OSes, etc :)

File: lib/shared/store.js (hunk 7 above)
Comment (Shou) [reply lib/shared/store.js]: The first is ensuring the drive should be auto-selected, the other is making sure the selected drives exist since this is an update to the available drives

File: lib/shared/models/available-drives.js [hunk 14]
```diff
@@ -68,3 +68,20 @@ exports.setDrives = (drives) => {
...
+/**
+ * @summary Find a drive by its device field
+ * @function
+ * @public
+ *
+ * @param {String} driveDevice - drive device identifier
+ * @returns {Object|undefined} drive object, or undefined
+ *
+ * @example
+ * const drive = availableDrives.findnDriveByDevice('/dev/sdb')
```
Comment (jhermsmeier) [inline lib/shared/models/available-drives.js]: Typo: findnDriveByDevice -> findDriveByDevice

**Comment by jhermsmeier** [comment]
Given the number of helper functions that are concerned with finding a drive by its "device" property; maybe we should consolidate them, and put them somewhere accessible to all consuming code?

File: lib/shared/models/selection-state.js [hunk 15]
```diff
@@ -272,76 +312,105 @@ exports.hasImage = () => {
...
  *
- * @param {Object} options - options
- * @param {Boolean} [options.preserveImage] - preserve image
- *
  * @example
- * selectionState.clear();
+ * selectionState.unselectAllDrives()
+ */
+exports.unselectAllDrives = () => {
+  _.map(exports.getSelectedDevices(), exports.unselectDrive)
```
Comment (jviotti) [inline lib/shared/models/selection-state.js]: Should this be `_.each`?

**Comment by Shou** [comment]
@jhermsmeier I looked into replacing all the finding with one generic helper function, but it would essentially have to be as generic as `_.find` itself, so instead I opted to just use that – the complexity is really in one state getter pattern used in the store so instead I made a helper function for that.

Once https://github.com/resin-io/etcher/pull/2043 is in, this PR is good for reviews.

File: lib/shared/store.js [hunk 16]
```diff
@@ -167,20 +163,20 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
       }
 
       // Convert object instances to plain objects
-      action.data = JSON.parse(JSON.stringify(action.data))
+      const drives = JSON.parse(JSON.stringify(action.data))
```
Comment (lurch) [inline lib/shared/store.js]: This might be another stupid question, but is there a more efficient way of doing this than transitioning to and from JSON?

File: lib/shared/store.js [hunk 17]
```diff
@@ -198,24 +194,35 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
...
-          type: ACTIONS.DESELECT_DRIVE
+          type: ACTIONS.DESELECT_DRIVE,
+          data: drive.device
         })
       }
 
-      return newState
+      const selectedDevices = newState.getIn([ 'selection', 'devices' ]).toJS()
+
+      return _.reduce(selectedDevices, (accState, device) => {
```
Comment (lurch) [inline lib/shared/store.js]: Could you add a code-comment explaining what this is doing?
I _think_ it's deselecting all drives? Would it be easier to just have a `ACTIONS.DESELECT_ALL_DRIVES` ? (which you could do just once _before_ testing the number of drives, so that you don't have two separate code-paths (as you have now) each doing `ACTIONS.DESELECT_DRIVE`)

File: lib/shared/store.js [hunk 18]
```diff
@@ -305,40 +312,44 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
     }
 
     case ACTIONS.SELECT_DRIVE: {
-      if (!action.data) {
+      const device = action.data
```
Comment (lurch) [inline lib/shared/store.js]: Sorry to keep banging on about consistency (but hey at least I'm consistent :wink: ), but on line 166 you have `const drives` but here you have `const device` :confused:

File: lib/shared/store.js [hunk 19]
```diff
@@ -432,24 +443,39 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
         })
       }
 
-      const selectedDrive = findDrive(state, state.getIn([ 'selection', 'drive' ]))
+      const selectedDevices = state.getIn([ 'selection', 'devices' ])
```
Comment (lurch) [inline lib/shared/store.js]: Line 350 (and line 475) has `const selectionDevices` but here you have `const selectedDevices`
(IMHO the latter sounds better)

File: lib/shared/store.js [hunk 20]
```diff
@@ -432,24 +443,39 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
...
 
-        return state
-      }).setIn([ 'selection', 'image' ], Immutable.fromJS(action.data))
+        return accState
+      }, state).setIn([ 'selection', 'image' ], Immutable.fromJS(action.data))
     }
 
     case ACTIONS.DESELECT_DRIVE: {
-      return state.deleteIn([ 'selection', 'drive' ])
+      if (!action.data) {
```
Comment (lurch) [inline lib/shared/store.js]: For consistency with the other actions, you could also assign this `action.data` to a const variable?

Comment (jhermsmeier) [reply lib/shared/store.js]: Let's not assign stuff to vars just for the sake of it; object property access is just fine, and should be preferred, unless too unweildy

File: lib/shared/store.js (hunk 16 above)
Comment (jhermsmeier) [reply lib/shared/store.js]: @Shou why is this object being copied again? I remember there was some reason for it, but since it's not in the comment...

File: lib/shared/store.js (hunk 20 above)
Comment (lurch) [reply lib/shared/store.js]: I dunno if the same is true in JavaScript-land, but I remember reading that in Python when you have performance-critical code, doing
```
foo = bar.thing
do_something(foo)
do_something_else(foo)
foo.wibble()
```
can be slightly faster than
```
do_something(bar.thing)
do_something_else(bar.thing)
bar.thing.wibble()
```
because it then does the property-lookup on `bar` just once rather than multiple times. \*shrug\*

> imo the `const something = action.data` should be removed in favor of accessing the props directly

Even if it makes the code less readable? \*shrug\* again ;)

File: lib/gui/app/components/drive-selector/controllers/drive-selector.js [hunk 21]
```diff
@@ -118,6 +118,7 @@ module.exports = function (
           previouslySelected: selectionState.isCurrentDrive(drive.device)
         })
 
+        selectionState.deselectOtherDrives(drive.device)
```
[the remark this answers is not in the document]
Comment (Shou) [reply lib/gui/app/components/drive-selector/controllers/drive-selector.js]: Given that this function isn't meant to be kept when we fully enable multi-selection for multi-writes it isn't something we need to worry about for now

File: lib/shared/models/selection-state.js [hunk 22]
```diff
@@ -82,19 +106,34 @@ exports.selectImage = (image) => {
...
- * @summary Get drive
+ * @summary Get all selected drives' devices
+ * @function
+ * @public
+ *
+ * @returns {String[]} selected drives' devices
+ *
+ * @example
+ * for (driveDevice of selectionState.getSelectedDevices()) {
+ *   console.log(driveDevice)
```
[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/models/selection-state.js]: One specifies what it returns (drive device field) and the other specifies the state it affects (drives selected), but I guess we could clarify by renaming it `deselectOtherDrivesByDevice` although that is a bit verbose and I think the `@param` field documents what it selects a drive by quite well already.

File: lib/shared/models/selection-state.js [hunk 23]
```diff
@@ -333,5 +387,29 @@ exports.isCurrentDrive = (driveDevice) => {
...
+ * @returns {Boolean}
+ *
+ * @example
+ * const isSelected = selectionState.isSelectedDrive('/dev/sdb')
+ *
+ * if (isSelected) {
+ *   selectionState.deselectDrive(driveDevice)
+ * }
+ */
+exports.isSelectedDrive = (driveDevice) => {
```
[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/models/selection-state.js]: I'm curious, what makes one preferred over the other?

File: lib/shared/store.js (hunk 16 above)
Comment (Shou) [reply lib/shared/store.js]: We tried using `Object.assign` once but it didn't fulfil the job completely – @jviotti knows the specifics. Maybe we can combine `Object.assign` with a function that takes care of the edge-case you found @jviotti? I imagine it'll be a lot faster than JSON conversion. We can do this in another PR though!

File: lib/shared/store.js (hunk 18 above)
Comment (Shou) [reply lib/shared/store.js]: `action.data` is used in a polymorphic manner and here it's a device field `String`, whereas on L166 it's a list of drives: `Array<Object>`. We could potentially change the field used in `action` and is definitely worth looking into in another issue/PR.

File: lib/shared/store.js (hunk 20 above)
Comment (Shou) [reply lib/shared/store.js]: The reason for the `const`s above is because it makes code in the same block easier to write (object shorthand syntax), whereas here it wouldn't help as much, but the discussion of `const` performance is worth continuing in an issue.

File: lib/shared/models/selection-state.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (lurch) [reply lib/shared/models/selection-state.js]: Maybe it's just me, but IMHO "isSelectedDrive" sounds like "is this the (one and only) selected drive", whereas "isDriveSelected" sounds like "is this drive selected (while other drives might be selected at the same time too)". IYSWIM :wink:

File: lib/shared/store.js (hunk 18 above)
Comment (lurch) [reply lib/shared/store.js]: Right. Can you see why I'm getting confused between 'drives' and 'devices' though? :slightly_smiling_face: 
(perhaps more verbose comments is all that's needed to clear things up?)

File: lib/shared/models/selection-state.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (lurch) [reply lib/shared/models/selection-state.js]: Maybe I'm just being a bit dim in grokking this code as I've not worked on it myself, but perhaps verbosity is a good thing? \*shrug\*

File: lib/shared/store.js (hunk 19 above)
Comment (Shou) [reply lib/shared/store.js]: I've renamed them `selectedDevices` now

File: lib/shared/store.js (hunk 17 above)
Comment (Shou) [reply lib/shared/store.js]: It's deselecting drives that don't exist in `availableDrives` anymore – I've added a few comments explaining that block now.

File: lib/shared/store.js [hunk 24]
```diff
@@ -122,25 +124,19 @@ const ACTIONS = _.fromPairs(_.map([
...
  *
- * @param {Object} state - application state
- * @param {String} device - drive device
- * @returns {(Object|Undefined)} drive
+ * @param {Object} state - state object
+ * @returns {Object} new state
  *
  * @example
- * const drive = findDrive(state, '/dev/disk2');
+ * utils.getDriveByDevice(getAvailableDrives(state), '/dev/sda')
```
[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/store.js]: Done

File: lib/shared/models/selection-state.js [hunk 25]
```diff
@@ -82,19 +106,34 @@ exports.selectImage = (image) => {
...
+/**
+ * @summary Get the head of the list of selected drives
  * @function
  * @public
  *
  * @returns {Object} drive
  *
  * @example
- * const drive = selectionState.getDrive();
+ * const drive = selectionState.getCurrentDrive();
```
[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/models/selection-state.js]: Done

File: lib/shared/models/selection-state.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/models/selection-state.js]: I've improved the docs here now

File: lib/shared/models/selection-state.js [hunk 26]
```diff
@@ -48,13 +48,37 @@ exports.selectDrive = (driveDevice) => {
...
+ * @description
+ * This is a temporary function during the transition to multi-writes,
+ * remove this and its uses when multi-selection should become user-facing.
+ *
+ * @param {String} driveDevice - drive device identifier
+ *
+ * @example
+ * selectionState.deselectOtherDrives('/dev/disk2')
+ */
+exports.deselectOtherDrives = (driveDevice) => {
```
[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/models/selection-state.js]: Done! I've also added tests just in case, because I don't trust this function anymore

File: lib/shared/store.js (hunk 18 above)
Comment (Shou) [reply lib/shared/store.js]: I wish we had TS here and that'd clarify it in the type :P

File: lib/shared/models/selection-state.js [hunk 27]
```diff
@@ -65,15 +65,19 @@ exports.toggleDrive = (driveDevice) => {
...
+ * > [ '/dev/disk1', '/dev/disk2', '/dev/disk3' ]
  * selectionState.deselectOtherDrives('/dev/disk2')
+ * console.log(selectionState.getSelectedDevices())
+ * > [ '/dev/disk2' ]
  */
-exports.deselectOtherDrives = (driveDevice) => {
-  if (exports.isSelectedDrive(driveDevice)) {
-    exports.deselectAllDrives()
-    exports.selectDrive(driveDevice)
+exports.deselectOtherDrives = (device) => {
```
Comment (lurch) [inline lib/shared/models/selection-state.js]: Umm, perhaps I'm missing something, but why does the `@example` for `deselectOtherDrives` show examples of `getSelectedDevices` ?

Comment (Shou) [reply lib/shared/models/selection-state.js]: It's to clarify what exactly it does. Do you think it's too verbose?

File: lib/shared/models/selection-state.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/models/selection-state.js]: That makes sense, I've changed it to `isDriveSelected` now

File: tests/shared/models/selection-state.spec.js [hunk 28]
```diff
@@ -225,6 +225,18 @@ describe('Model: selectionState', function () {
...
+
+    describe('.deselectOtherDrives()', function () {
+      it('should deselect other drives', function () {
+        selectionState.deselectOtherDrives('/dev/disk1')
+        m.chai.expect(selectionState.getSelectedDevices()).to.not.include.members([ '/dev/disk2' ])
+      })
+
+      it('should not remove the specified drive', function () {
+        selectionState.deselectOtherDrives('/dev/sdb')
+        m.chai.expect(selectionState.getSelectedDevices()).to.deep.equal([ this.drives[0].device ])
```
Comment (lurch) [inline tests/shared/models/selection-state.spec.js]: You've already got `/dev/sdb` on the line above, so it might be more logical for this line to say:
```
m.chai.expect(selectionState.getSelectedDevices()).to.deep.equal([ '/dev/sdb' ])
```
?

File: lib/shared/models/selection-state.js (hunk 27 above)
Comment (lurch) [reply lib/shared/models/selection-state.js]: Sorry, my fault. I was misreading the diff and hadn't spotted that you'd still got `deselectOtherDrives` in the middle of the example :laughing: 
image

File: tests/shared/models/selection-state.spec.js (hunk 28 above)
Comment (Shou) [reply tests/shared/models/selection-state.spec.js]: I've opted for the other route as it looks more consistent with the tests above – use `this.drives` more.

File: lib/shared/store.js [hunk 29]
```diff
@@ -229,6 +231,8 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
     }
 
     case ACTIONS.SET_FLASH_STATE: {
+      // action.data : Object
```
Comment (lurch) [inline lib/shared/store.js]: Nice, but given there's now so many lines that say `// action.data : Object`, I wonder if these too should be a bit more verbose e.g. `// action.data : Object (flash state)` ? (or `// action.data : flash state Object`)
Or perhaps that's unnecessary as that info is implicit from the `case` line (e.g. `ACTIONS.SET_FLASH_STATE`) ?

Comment (Shou) [reply lib/shared/store.js]: Done that now (and also the linter requires sentence-case in comments, so that's fixed too)

File: lib/shared/store.js [hunk 30]
```diff
@@ -157,7 +157,7 @@ const getAvailableDrives = (state) => {
 const storeReducer = (state = DEFAULT_STATE, action) => {
   switch (action.type) {
     case ACTIONS.SET_AVAILABLE_DRIVES: {
-      // action.data : Array<DriveObject>
+      // Type: action.data : Array<DriveObject>
```
Comment (lurch) [inline lib/shared/store.js]: Is this some 'convention' I'm not aware of? If not, I think having two colons on the one line looks a bit strange?

File: lib/shared/store.js (hunk 16 above)
Comment (jviotti) [reply lib/shared/store.js]: I don't remember the details anymore, sorry :(

Comment (jhermsmeier) [reply lib/shared/store.js]: Yes, good point, let's figure that out in a different PR

File: tests/shared/models/selection-state.spec.js (hunk 28 above)
Comment (lurch) [reply tests/shared/models/selection-state.spec.js]: Sweet. And I've just realised that `this.drives` has both `/dev/sdb` (i.e. Linux-style device strings) and `/dev/disk2` (i.e. OSX-style device strings) which is slightly confusing! :laughing: (but that's obviously outside the scope of this PR)

File: lib/gui/app/components/drive-selector/templates/drive-selector-modal.tpl.html [hunk 31]
```diff
@@ -38,7 +38,7 @@ <h4 class="list-group-item-heading">{{ drive.description }}
         </div>
         <span class="list-group-item-section tick tick--success"
           ng-show="modal.constraints.isDriveValid(drive, modal.state.getImage())"
-          ng-disabled="!modal.state.isSelectedDrive(drive.device)"></span>
+          ng-disabled="!modal.state.isDriveSelected(drive.device)"></span>
```
Comment (lurch) [inline lib/gui/app/components/drive-selector/templates/drive-selector-modal.tpl.html]: :+1:  `isDriveSelected` also looks nicer next to `isDriveValid` :)

File: lib/shared/store.js (hunk 30 above)
Comment (Shou) [reply lib/shared/store.js]: `Type:` isn't really a convention AFAIK, it's just a nod to the user reading it, but I can see why it would confuse and I'm not sure how to resolve this one now. `Type(action.data : Array<DriveObject>)`? `Type - action.data : Array<DriveObject>`?

File: lib/shared/models/selection-state.js [hunk 32]
```diff
@@ -48,13 +48,41 @@ exports.selectDrive = (driveDevice) => {
...
+ * @summary Deselect all other drives and keep the current drive's status
+ * @function
+ * @public
+ * @deprecated
+ *
+ * @description
+ * This is a temporary function during the transition to multi-writes,
+ * remove this and its uses when multi-selection should become user-facing.
+ *
+ * @param {String} device - drive device identifier
```
Comment (lurch) [inline lib/shared/models/selection-state.js]: Sorry to be a pain, but here you've got `@param {String} device - drive device identifier` and on e.g. line 406 you've got `@param {String} driveDevice - drive device identifier` - perhaps it makes sense to choose one or the other and be consistent?

Comment (Shou) [reply lib/shared/models/selection-state.js]: No worries, you're right, I made that change to leverage object shorthand syntax but now that I've ended up not using it, it can be `driveDevice` with no drawbacks. It's done now.

File: lib/shared/store.js (hunk 30 above)
Comment (lurch) [reply lib/shared/store.js]: Anything wrong with just `action.data : Array<DriveObject>` - should be fairly self-evident what it means?

File: lib/shared/store.js [hunk 33]
```diff
@@ -122,25 +124,20 @@ const ACTIONS = _.fromPairs(_.map([
...
  */
-const findDrive = (state, device) => {
-  /* eslint-disable lodash/prefer-lodash-method */
-
-  return state.get('availableDrives').find((drive) => {
-    return drive.get('device') === device
-  })
-
-  /* eslint-enable lodash/prefer-lodash-method */
+const getAvailableDrives = (state) => {
```
Comment (lurch) [inline lib/shared/store.js]: I dunno how all the store-related classes are supposed to interact, but is this a "duplicate" of https://github.com/resin-io/etcher/blob/master/lib/shared/models/available-drives.js#L68 ?

Comment (Shou) [reply lib/shared/store.js]: Yup, but `availableDrives` relies on `store` so we can't expose that, and I don't think we want to expose anything else from the `store` either

File: tests/shared/models/selection-state.spec.js (hunk 28 above)
Comment (lurch) [reply tests/shared/models/selection-state.spec.js]: Or maybe it _is_ within the scope if this PR? I was only looking at a single commit before and didn't realise that this PR was adding the entirety of the `this.drives` object.

File: lib/shared/models/selection-state.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (lurch) [reply lib/shared/models/selection-state.js]: I _think_ I'm starting to finally understand it all now - and the extra comments help, thank you :)

So we have a `drive` which is an Object, and a `driveDevice` which is a String, and each `drive` Object has a `device` String property (amongst others) which corresponds to the `driveDevice` strings. And "availableDrives" (in the `store`) is a list of these `drive` Objects, and "selectedDevices" (also in the `store`) is a list of `driveDevice` strings. And there's various methods to manipulate/translate these, e.g. getting the `drive` Object(s) that correspond to a particular set of `driveDevice` strings, and vice-versa.

Why is the "availableDrives" stored as an array and not as an object mapping `driveDevice` String -> `drive` Object? (seems like that'd make a lot of the manipulations / translations a bit easier?)

And listing the drive/device-related functions in `selection-state.js` (using a modified notation) we have:
```
selectDrive(String driveDevice)
toggleDrive(String driveDevice)
deselectOtherDrives(String driveDevice)
String[] driveDevices getSelectedDevices()
Object drive getCurrentDrive()
Boolean hasDrive()
deselectDrive(String driveDevice)
deselectAllDrives()
Boolean isCurrentDrive(String driveDevice)
Boolean isDriveSelected(String driveDevice)
```
...so I think the reason I was getting so confused is that we have lots of functions named `xxxDrive` which actually take a `driveDevice` as a parameter, another function named `xxxDrive()` (`getCurrentDrive`) that returns a `drive` Object, and only one function-name named `xxxDevice` (`getSelectedDevices`) that returns a list of `driveDevice`s. :upside_down_face:  (and there's also the asymmetry of `getCurrentDrive` and `isCurrentDrive` returning / taking different types, even though they're named so similarly!)


And again it's something well outside the scope of this particular PR, but I wonder if (in future) it might make sense to split `selection-state.js` up into `drive-selection-state.js` and `image-selection-state.js` ?

File: lib/shared/store.js [hunk 34]
```diff
@@ -122,25 +124,20 @@ const ACTIONS = _.fromPairs(_.map([
...
+ * @summary Get available drives from the state
  * @function
- * @private
+ * @public
  *
- * @param {Object} state - application state
- * @param {String} device - drive device
- * @returns {(Object|Undefined)} drive
+ * @param {Object} state - state object
+ * @returns {Object} new state
```
Comment (lurch) [inline lib/shared/store.js]: The `@returns` here is still wrong, isn't it?

File: lib/gui/app/pages/main/templates/main.tpl.html [hunk 35]
```diff
@@ -66,11 +66,11 @@
               'text-disabled': main.shouldDriveStepBeDisabled()
             }">
             <span class="drive-step step-name"
-              uib-tooltip="{{ main.selection.getDrive().description }} ({{ main.selection.getDrive().displayName }})">
+              uib-tooltip="{{ main.selection.getCurrentDrive().description }} ({{ main.selection.getCurrentDrive().displayName }})">
               <!-- middleEllipses errors on undefined, therefore fallback to empty string -->
-              {{ (main.selection.getDrive().description || "") | middleEllipses:11 }}
+              {{ (main.selection.getCurrentDrive().description || "") | middleEllipses:11 }}
```
Comment (lurch) [inline lib/gui/app/pages/main/templates/main.tpl.html]: In `selection-state.js` there's a bunch of "helper methods" like `getImageName` and `getImageSize`. Would it also make sense for there to be "helper methods" like `getCurrentDriveDescription()` and `getCurrentDriveSize()` ?

File: lib/shared/models/selection-state.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/models/selection-state.js]: Using a dictionary/map/object instead of a list is a good idea as you should never have two identical mount points – I assume it's faster too. @jhermsmeier what do you think?

As for splitting it, that's definitely something I think we'll go towards once we get drives and HTTP as sources.

[the remark this answers is not in the document]
Comment (Shou) [reply lib/shared/models/selection-state.js]: And yeah, you've got it right, here :)

[the remark this answers is not in the document]
Comment (jhermsmeier) [reply lib/shared/models/selection-state.js]: @Shou yeah, I think it's a good idea – that will very much simplify things I reckon; performance wise the impact is negligible compared to everything else going on, as this code isn't called very often (i.e. not 1000+ op/s)

[the remark this answers is not in the document]
Comment (jhermsmeier) [reply lib/shared/models/selection-state.js]: Do you wanna do that in a follow up PR? This one has become quite the one to navigate :D

**Comment by lurch** [comment]
Sorry all my comments made this PR drag on for so long, but hopefully the code is much better for it :slightly_smiling_face:
